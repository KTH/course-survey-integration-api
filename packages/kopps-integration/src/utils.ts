import { InvalidLadokUidError, SyllabusNotFoundError } from "./errors";
import type {
  KoppsCourseDetailedInformation,
  KoppsCourseRoundInfo,
  KoppsSyllabus,
} from "./types";

const PERIODS = ["P0", "P1", "P2", "P3", "P4", "P5"] as const;
type Periods = (typeof PERIODS)[number];

/**
 * Return the periods in a round object. Periods are sorted from oldest to
 * latest by the term where the period belongs.
 */
export function getPeriods(roundObject: KoppsCourseRoundInfo) {
  const periods: Periods[] = [];
  const sortedTerms = roundObject.round.courseRoundTerms.sort(
    (t1, t2) => t1.term.term - t2.term.term,
  );

  for (const c of sortedTerms) {
    // Add the number to the "periods" array if the number of credits is higher than 0
    for (const p of PERIODS) {
      const numberOfCredits = c[`credits${p}`];
      if (numberOfCredits && numberOfCredits > 0) {
        periods.push(p);
      }
    }
  }

  return periods;
}

/** Return the course round information given a Ladok UID */
export function getCourseRoundInformation(
  detailedInformation: KoppsCourseDetailedInformation,
  ladokUid: string,
): KoppsCourseRoundInfo {
  const roundInformation = detailedInformation.roundInfos.find(
    (r) => r.round.ladokUID === ladokUid,
  )!;

  if (!roundInformation) {
    throw new InvalidLadokUidError(ladokUid);
  }

  return roundInformation;
}

/** Return the most recent syllabus that is valid in the given term */
export function getSyllabus(
  allSyllabus: KoppsSyllabus[],
  term: number,
): KoppsSyllabus {
  // Filter-out the invalid ones
  const validSyllabus = allSyllabus.filter((s) => s.validFromTerm.term <= term);

  if (validSyllabus.length === 0) {
    throw new SyllabusNotFoundError(term);
  }

  // Take the latest
  return validSyllabus.sort(
    (s1, s2) => s2.validFromTerm.term - s1.validFromTerm.term,
  )[0];
}
