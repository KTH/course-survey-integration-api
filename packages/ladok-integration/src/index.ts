import {
  getKursinstans,
  getKurstillfalle,
  getKurstillfallesdeltagande,
  getOrganisation,
  getStudiestruktur,
  getUtbildningsinstans,
  getUtbildningstillfalle,
} from "./api";
import {
  diffTerms,
  findStudiestruktur,
  getGradingScheme,
  getTermFromDate,
  isDoktorsavhandling,
  isKurspaketering,
  isUppdragsutbildning,
  isUtbytesinstans,
  isUtbytesstudent,
  parseOrganisation,
} from "./utils";

export { getGradingScheme } from "./utils";

export type ProgramParticipation =
  | undefined
  | {
      code: string;
      name: { sv: string; en: string };
      startTerm: string;
      studyYear: number;
      specialization:
        | undefined
        | {
            code: string;
            name: { sv: string; en: string };
          };
    };

type TMultiLang = {
  sv?: string;
  en?: string;
};

type TGradingScheme = {
  code: string;
  grades: Array<{
    validFinalGrade: boolean;
    code: string;
  }>;
};

export type TGetCourseRoundInformation = {
  name: TMultiLang;
  courseCode: string;
  organisation: {
    code: string;
  };
  organisationUnit: {
    code: string;
    name: TMultiLang;
  };
  courseInstanceCode: string;
  startDate: string;
  endDate: string;
  credits: number;
  modules: Array<{
    code: string;
    name: TMultiLang;
    credits: number;
    gradingScheme: TGradingScheme;
  }>;
  gradingScheme: TGradingScheme;
};

export type TGetEduInstance = {
  ladokUID: string;
  isCoursePackage: boolean;
  isDoctoralThesis: boolean;
  isExchangeCourse: boolean;
  isExchangeStudent: boolean;
  isIndustrialEdu: boolean;
}

export type TGetCourseRoundLanguage = {
  language: "sv" | "en";
}

export async function getCourseRoundLanguage(
  utbildningstillfalleUid: string,
): Promise<TGetCourseRoundLanguage> {
  const utbildningstillfalle = await getUtbildningstillfalle(utbildningstillfalleUid);
  const langAttr = utbildningstillfalle.Attributvarden.find(a => a.Namn === "utbildning.attribut.undervisningssprak");
  return {
    language: langAttr?.Varde === "1" ? "sv" : "en",
  };
}

/** Get course round information from its Ladok UID */
export async function getCourseRoundInformation(
  ladokUid: string,
): Promise<TGetCourseRoundInformation> {
  const kurstillfalle = await getKurstillfalle(ladokUid);
  const kursinstans = await getKursinstans(kurstillfalle.UtbildningsinstansUID);
  const organisation = parseOrganisation(
    await getOrganisation(kursinstans.OrganisationUID),
  );

  return {
    name: kurstillfalle.Benamning,
    courseCode: kurstillfalle.Utbildningskod,
    courseInstanceCode: kurstillfalle.Kurstillfalleskod,
    startDate: kurstillfalle.Startdatum,
    endDate: kurstillfalle.Slutdatum,
    organisation: organisation.school,
    organisationUnit: organisation.department,
    credits: kurstillfalle.Omfattning,
    modules: kurstillfalle.IngaendeMoment.map((m) => ({
      code: m.Utbildningskod,
      name: m.Benamning,
      credits: m.Omfattning,
      gradingScheme: getGradingScheme(m.BetygsskalaID),
    })),

    // TODO: Extract the information from kurstillfalle.BetygsskalaID
    gradingScheme: getGradingScheme(kurstillfalle.BetygsskalaID),
  };
}

export async function getEduInstance(ladokUid: string): Promise<TGetEduInstance> {
  // Course packages don't have kurstillfalle, otherwise that would have been a
  // nicer endpoint.
  const utbtillfalle = await getUtbildningstillfalle(ladokUid);
  const utbinstans = await getUtbildningsinstans(utbtillfalle.UtbildningsinstansUID);
  const isCoursePackage = isKurspaketering(utbinstans);
  const isDoctoralThesis = isDoktorsavhandling(utbinstans);
  const isExchangeCourse = isUtbytesinstans(utbinstans);
  const isExchangeStudent = isUtbytesstudent(utbinstans);
  const isIndustrialEdu = isUppdragsutbildning(utbinstans);

  return {
    ladokUID: ladokUid,
    isCoursePackage,
    isDoctoralThesis,
    isExchangeCourse,
    isExchangeStudent,
    isIndustrialEdu,
  };
}

/**
 * Get all programs that include the course round where a student is participating
 */
export async function getProgramParticipation(
  studentUID: string,
  courseRoundUID: string,
): Promise<ProgramParticipation> {
  const d = await getKurstillfallesdeltagande(studentUID);
  const d1 = d.Tillfallesdeltaganden.find(
    (t) => t.Utbildningsinformation.UtbildningstillfalleUID === courseRoundUID,
  );

  if (!d1 || !d1.Studiestrukturreferens) {
    // OK. The course round does not belong to any program
    return undefined;
  }

  if (d1.Utbildningsinformation.Utbildningskod === "UTVTS") {
    // This is part of a study exchange, not a specific program
    // FIX https://kth-se.atlassian.net/browse/FOE-418
    return undefined;
  }

  const s = await getStudiestruktur(studentUID);
  const arr = findStudiestruktur(d1.Studiestrukturreferens, s.Studiestrukturer);

  if (arr.length === 0) {
    throw new Error(
      `The student [${studentUID}] participates in course round [${courseRoundUID}] which has a "strukturreferens" [${d1.Studiestrukturreferens}], but there is no program with such "studiestrukturreferens"`,
    );
  }

  const courseRoundStartTerm = getTermFromDate(
    d1.Utbildningsinformation.Studieperiod.Startdatum,
  );
  const program = arr[0];
  const programStartTerm = getTermFromDate(
    program.Utbildningsinformation.Studieperiod.Startdatum,
  );
  const diff = diffTerms(courseRoundStartTerm, programStartTerm);
  const studyYear = Math.floor(diff / 2) + 1;
  const programData = {
    code: program.Utbildningsinformation.Utbildningskod,
    name: program.Utbildningsinformation.Benamning,
    startTerm: programStartTerm,
    studyYear,
    specialization: undefined,
  };

  if (arr.length === 1) {
    return programData;
  }

  const specialization = arr[1];
  const specializationData = {
    code: specialization.Utbildningsinformation.Utbildningskod,
    name: specialization.Utbildningsinformation.Benamning,
  };

  if (arr.length === 2) {
    return {
      ...programData,
      specialization: specializationData,
    };
  }

  throw new Error(
    `The student [${studentUID}] participates in course round [${courseRoundUID}], which is related to a "kurspaketering" with ${arr.length} levels (expected a maximum of two: program and specialization)`,
  );
}
