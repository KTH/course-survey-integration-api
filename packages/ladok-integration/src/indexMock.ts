import { ProgramParticipation, TGetCourseRoundInformation, TGetCourseRoundLanguage } from ".";

// This doesn't require mocking
export { getGradingScheme } from "../src";

const _mockedValues: Record<
  string,
  TGetCourseRoundInformation | ProgramParticipation | TGetCourseRoundLanguage
> = {};

export class LadokIntegrationMock {
  static getCourseRoundLanguage(
    ladokUid: string,
    params: TGetCourseRoundLanguage,
  ) {
    if (_mockedValues[ladokUid])
      throw new Error(`Mock value already registered for ${ladokUid}`);

    _mockedValues[ladokUid] = params;
  }

  static getCourseRoundInformation(
    ladokUid: string,
    params: TGetCourseRoundInformation,
  ) {
    if (_mockedValues[ladokUid])
      throw new Error(`Mock value already registered for ${ladokUid}`);

    _mockedValues[ladokUid] = params;
  }

  static getProgramParticipation(
    studentUID: string,
    courseRoundUID: string,
    params: ProgramParticipation,
  ) {
    const key = [studentUID, courseRoundUID].join("-");
    if (_mockedValues[key])
      throw new Error(`Mock value already registered for ${key}`);

    _mockedValues[key] = params;
  }
}

export async function getCourseRoundLanguage(utbildningstillfalleUid: string) {
  return _mockedValues[utbildningstillfalleUid];
}


export async function getCourseRoundInformation(ladokUid: string) {
  return _mockedValues[ladokUid];
}

export async function getProgramParticipation(
  studentUID: string,
  courseRoundUID: string,
) {
  return _mockedValues[[studentUID, courseRoundUID].join("-")];
}
