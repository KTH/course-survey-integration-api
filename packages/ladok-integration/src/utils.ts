import type { z } from "zod";
import { LadokOrganisation } from "./types";
import assert from "node:assert";

type LadokOrganisation = z.infer<typeof LadokOrganisation>;

/**
 * This information is obtained from Ladok
 * But hardcoded for eficiency reasons
 */
export function getGradingScheme(id: number) {
  // AF
  if (id === 131657) {
    return {
      code: "AF",
      grades: [
        { validFinalGrade: true, code: "A" },
        { validFinalGrade: true, code: "B" },
        { validFinalGrade: true, code: "C" },
        { validFinalGrade: true, code: "D" },
        { validFinalGrade: true, code: "E" },
        { validFinalGrade: false, code: "FX" },
        { validFinalGrade: false, code: "F" },
      ],
    };
  }

  // PF
  if (id === 131656) {
    return {
      code: "PF",
      grades: [
        { validFinalGrade: true, code: "P" },
        { validFinalGrade: false, code: "F" },
      ],
    };
  }

  throw new Error(`Unknown grading scheme id [${id}]`);
}

export function parseOrganisation(organisation: LadokOrganisation) {
  // Both "Benamning.sv" and "Benamning.en" are formatted as
  // "SCI/Matematik" where "SCI" is the School code and should be the same
  // in both languages
  const [codeSv, departmentSv] = organisation.Benamning.sv.split("/");
  const [codeEn, departmentEn] = organisation.Benamning.en.split("/");

  // Verifying the length of the arrays
  if (!departmentEn) {
    throw new Error(
      `Incompatible format. Benamning.en must be "School code/Institution name". Obtained: [${organisation.Benamning.en}]`,
    );
  }

  if (!departmentSv) {
    throw new Error(
      `Incompatible format. Benamning.sv must be "School code/Institution name". Obtained: [${organisation.Benamning.sv}]`,
    );
  }

  if (codeSv !== codeEn) {
    throw new Error(
      `School code in Swedish [${codeSv}] does not match with code in English [${codeEn}]`,
    );
  }

  return {
    school: {
      code: codeSv,
    },
    department: {
      code: organisation.Organisationskod,
      name: {
        en: departmentEn,
        sv: departmentSv,
      },
    },
  };
}

/**
 * Get the year and term from a date
 * @param {string} date  Date formatted as `YYYY-MM-DD`
 * @return a string "XXXXY" where XXXX is a year and Y is the term (1 for spring, 2 for autumn)
 */
export function getTermFromDate(date: string) {
  const REGEX = /(\d\d\d\d)\-(\d\d)\-(\d\d)/;
  const m = date.match(REGEX);

  assert(m !== null, "Error!!1");

  const year = m[1];
  const month = parseInt(m[2], 10);

  if (month <= 6) {
    return `${year}1`;
  } else {
    return `${year}2`;
  }
}

/** Get the difference between two terms */
export function diffTerms(t1: string, t2: string) {
  // 1. Number of years between the periods
  const yDiff = parseInt(t1.slice(0, 4), 10) - parseInt(t2.slice(0, 4), 10);

  // 2. If they are not the same period, add or substract 1
  const tDiff = parseInt(t1.slice(-1), 10) - parseInt(t2.slice(-1), 10);

  return yDiff * 2 + tDiff;
}

/**
 * Given the reference of a "Studiestruktur" and a tree with "Studiestruktur",
 * find the leaf in the structure with the reference
 *
 * Returns an array of the path from the root to the found leaf
 */
export function findStudiestruktur<T extends { Referens: string; Barn: T[] }>(
  ref: string,
  struktur: T[],
): Omit<T, "Barn">[] {
  for (const leaf of struktur) {
    const { Barn, ...other } = leaf;

    if (leaf.Referens === ref) {
      return [other];
    }

    if (leaf.Barn.length > 0) {
      const subStructure = findStudiestruktur(ref, leaf.Barn);

      if (subStructure.length > 0) {
        return [other, ...subStructure];
      }
    }
  }

  return [];
}
