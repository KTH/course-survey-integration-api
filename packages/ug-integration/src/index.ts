import { UgUser, checkGetUgCourseResponsibleAndTeachers } from "./types";
import { UGRestClient, UGRestClientError } from "./ugRestClient";
import { ZodError, ZodObject, ZodRawShape } from "zod";

const {
  OAUTH_SERVER_BASE_URI,
  UG_REST_API_BASEURL,
  UG_REST_API_CLIENT_ID,
  UG_REST_API_CLIENT_SECRET,
} = process.env;

const ugClient = new UGRestClient({
  authServerDiscoveryURI: OAUTH_SERVER_BASE_URI ?? '',
  resourceBaseURI: UG_REST_API_BASEURL ?? '',
  clientId: UG_REST_API_CLIENT_ID ?? '',
  clientSecret: UG_REST_API_CLIENT_SECRET ?? '',
});

export type TUgCourseResponsibleAndTeachers = [
  courseResponsible: string,
  courseTeachers: string[],
]

export async function getUgCourseResponsibleAndTeachers(courseCode: string, roundYear: string, roundCode: string | number): Promise<TUgCourseResponsibleAndTeachers | []> {
  const prefix = courseCode.slice(0, 2);
  const path = `edu.courses.${prefix}.${courseCode}.${roundYear}${roundCode}`

  // The course can have several parallel sections so we need to get all of them
  // Example: edu.courses.SF.SF1625.20222

  // TODO: I can't get the subgroups from the query so I can't iterate over the sections
  const { data, json, statusCode } =
    (await ugClient
      .get<any[]>(
        `groups?$filter=startswith(name,'${path}')` // &$expand=Subgroups
      )
      .catch(ugClientGetErrorHandler)) ?? <any>{};
  if (statusCode !== 200) {
    throw new Error(`UGRestClient: ${statusCode} ${data}`);
  }

  if (json === undefined) return [];

  const courseResonsible = json
    .filter((group: any) => group.name.endsWith('courseresponsible'))
    .reduce((val: string | undefined, curr: any) => val ?? curr.members.pop(), undefined);
  const courseTeachers = json
    .filter((group: any) => group.name.endsWith('teachers'))
    .reduce((val: string[], curr: any) => {
      // Add only unique values
      return [...val, ...curr.members.filter((m: any) => !val.includes(m))]
    }, []);    

  // This will throw if values are incorrect
  // TODO: Investigate if zod can be used here too
  // TODO: Improve error handling to level of ladok-integration
  checkGetUgCourseResponsibleAndTeachers([courseResonsible, courseTeachers]);

  return [
    courseResonsible,
    courseTeachers,
  ];
}

export type TUgUser = {
  email: string;
  kthid: string;
  givenName: string;
  surname: string;
}

export async function getUgUser(kthId: string |Â undefined): Promise<TUgUser | undefined>  {
  if (kthId === undefined) return;

  const { data, json, statusCode } =
    (await ugClient
      .get<TUgUser[]>(
        // `groups?$filter=name eq 'edu.courses.SF.SF1625.20222'`
        `users?$filter=kthid eq '${kthId}'`
      )
      .catch(ugClientGetErrorHandler)) ?? <any>{};

  if (json === undefined) return;

  const outp = json.pop();

  // Type-check with zod, throws error if values are incorrect
  // TODO: Improve error handling to level of ladok-integration
  UgUser.parse(outp)

  return outp;
}

function ugClientGetErrorHandler(err: any): never {
  if (err instanceof UGRestClientError) {
    throw err;
  }

  Error.captureStackTrace(err, ugClientGetErrorHandler);
  throw err;
}