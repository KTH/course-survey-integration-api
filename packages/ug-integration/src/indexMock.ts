import { TUgSchool, TUgUser } from ".";

const _mockedValues: {
  getUgMembers: Record<string, string[]>;
  getUgCourseResponsible: Record<string, string[]>;
  getUgCourseTeachers: Record<string, string[]>;
  getUgCourseExaminers: Record<string, string[]>;
  getUgUser: Record<string, TUgUser>;
  getUgSchool: Record<string, TUgSchool>;
  getUgUserByLadokId: Record<string, TUgUser>;
} = {
  getUgMembers: {},
  getUgCourseResponsible: {},
  getUgCourseTeachers: {},
  getUgCourseExaminers: {},
  getUgUser: {},
  getUgSchool: {},
  getUgUserByLadokId: {},
};

export class UgIntegrationMock {
  static getUgMembers(groupName: string, params: string[]) {
    const mocked = _mockedValues.getUgMembers;
    if (mocked[groupName])
      throw new Error(`Mock value already registered for ${groupName}`);

    return (mocked[groupName] = params);
  }

  static getUgCourseResponsible(
    courseCode: string,
    roundYear: string,
    roundCode: string,
    params: string[],
  ) {
    const mocked = _mockedValues.getUgCourseResponsible;
    const key = `${courseCode}-${roundYear}-${roundCode}`;
    if (mocked[key])
      throw new Error(`Mock value already registered for ${key}`);
    mocked[key] = params;
  }

  static getUgCourseTeachers(
    courseCode: string,
    roundYear: string,
    roundCode: string,
    params: string[],
  ) {
    const mocked = _mockedValues.getUgCourseTeachers;
    const key = `${courseCode}-${roundYear}-${roundCode}`;
    if (mocked[key])
      throw new Error(`Mock value already registered for ${key}`);
    mocked[key] = params;
  }

  static getUgCourseExaminers(courseCode: string, params: string[]) {
    const mocked = _mockedValues.getUgCourseExaminers;
    if (mocked[courseCode])
      throw new Error(`Mock value already registered for ${courseCode}`);
    mocked[courseCode] = params;
  }

  static getUgUser(kthId: string, params: TUgUser) {
    const mocked = _mockedValues.getUgUser;
    if (mocked[kthId])
      throw new Error(`Mock value already registered for ${kthId}`);

    mocked[kthId] = params;
  }

  static getUgSchool(schoolCode: string, params: TUgSchool) {
    const mocked = _mockedValues.getUgSchool;
    if (mocked[schoolCode])
      throw new Error(`Mock value already registered for ${schoolCode}`);

    mocked[schoolCode] = params;
  }

  static getUgUserByLadokId(ladokId: string, params: TUgUser) {
    const mocked = _mockedValues.getUgUserByLadokId;
    if (mocked[ladokId])
      throw new Error(`Mock value already registered for ${ladokId}`);

    mocked[ladokId] = params;
  }
}

export async function getUgMembers(groupName: string) {
  return _mockedValues.getUgMembers[groupName];
}

export async function getUgCourseResponsible(
  courseCode: string,
  roundYear: string,
  roundCode: string,
) {
  const key = `${courseCode}-${roundYear}-${roundCode}`;
  return _mockedValues.getUgCourseResponsible[key];
}

export async function getUgCourseTeachers(
  courseCode: string,
  roundYear: string,
  roundCode: string,
) {
  const key = `${courseCode}-${roundYear}-${roundCode}`;
  return _mockedValues.getUgCourseTeachers[key];
}

export async function getUgCourseExaminers(courseCode: string) {
  return _mockedValues.getUgCourseExaminers[courseCode];
}

export async function getUgUser(
  kthId: string | undefined,
): Promise<TUgUser | undefined> {
  if (kthId === undefined) return;
  return _mockedValues.getUgUser[kthId];
}

export async function getUgUserByLadokId(
  ladokId: string,
): Promise<TUgUser | undefined> {
  return _mockedValues.getUgUserByLadokId[ladokId];
}

export async function getUgSchool(
  schoolCode: string,
): Promise<TUgSchool | undefined> {
  return _mockedValues.getUgSchool[schoolCode];
}
