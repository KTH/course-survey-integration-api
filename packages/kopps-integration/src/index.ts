import { getCourseDetailedInformation, getCourseRoundSummary } from "./api";
import { getCourseRoundInformation, getPeriods, getSyllabus } from "./utils";

/** Get course round information from its Laodk UID */
export async function getCourseInformation(ladokUid: string) {
  const roundSummary = await getCourseRoundSummary(ladokUid);
  const detailedInformation = await getCourseDetailedInformation(
    roundSummary.course_code,
  );
  const roundInformation = getCourseRoundInformation(
    detailedInformation,
    ladokUid,
  );
  const term = roundInformation.round.startTerm.term;
  const syllabus = getSyllabus(
    detailedInformation.publicSyllabusVersions,
    term,
  );

  return {
    course: {
      name: {
        sv: detailedInformation.course.title,
        en: detailedInformation.course.titleOther,
      },
      code: detailedInformation.course.courseCode,
    },
    round: {
      /** Start year and term. Formatted as XXXXY */
      startTerm: term.toString(10),

      /** Deprecated unique identifier for course rounds (one-digit) */
      code: roundInformation.round.ladokRoundId,
      /**
       * Period information sorted from the "oldest" to the "latest".
       * Example: if the course round includes the periods 2023-P5, 2024-P1 and 2025-P2
       * the function will return [5, 1, 2] in that order
       */
      periods: getPeriods(roundInformation),
    },
    syllabus: {
      goals: syllabus.courseSyllabus.goals,
    },
  };
}
