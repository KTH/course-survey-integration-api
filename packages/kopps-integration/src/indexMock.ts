import { getCourseInformation as origGetCourseInformation } from ".";
import { CourseElectiveCondition } from "./types";
import { getPeriods } from "./utils";

type TParams = {
  title: string;
  titleOther: string;
  courseCode: string;
  periods: ReturnType<typeof getPeriods>;
  goals: string;
  startTerm: string;
  code: string;
  electiveConditions: CourseElectiveCondition[];
};

const _mockedValues: Record<string, TParams> = {};

export class KoppsIntegrationMock {
  static getCourseInformation(ladokUid: string, params: TParams) {
    if (_mockedValues[ladokUid])
      throw new Error(`Mock value already registered for ${ladokUid}`);

    _mockedValues[ladokUid] = params;
  }
}

export async function getCourseInformation(
  ladokUid: string,
): Promise<ReturnType<typeof origGetCourseInformation>> {
  const { title, titleOther, courseCode, periods, goals, startTerm, code, electiveConditions } =
    _mockedValues[ladokUid];
  return {
    course: {
      name: {
        sv: title,
        en: titleOther,
      },
      code: courseCode,
    },
    round: {
      startTerm,
      code,
      periods,
    },
    syllabus: {
      goals: goals,
    },
    electiveConditions,
  };
}
