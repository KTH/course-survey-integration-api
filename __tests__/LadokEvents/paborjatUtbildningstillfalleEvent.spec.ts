import { UgIntegrationMock } from "ug-integration/src/indexMock";
import event from "../../__fixtures__/events/01_studiedeltagande.PaborjatUtbildningstillfalleEvent_1.json";
import { handler } from "../../src/functions/ladok-events/paborjatUtbildningstillfalleEvent";
import { MockContext } from "./utils/mockContext";
import { MockDatabase } from "./utils/mockDatabase";
import { KoppsIntegrationMock } from "kopps-integration/src/indexMock";
import { LadokIntegrationMock } from "ladok-integration/src/indexMock";

KoppsIntegrationMock.getCourseInformation(
  event.message.UtbildningstillfalleUID,
  {
    title: "Programmeringsteknik för F",
    titleOther: "Programming in F",
    courseCode: "DD1321",
    periods: ["P1"],
    goals: "The course aims to give basic knowledge of programming in F#.",
  },
);

LadokIntegrationMock.getCourseRoundInformation(
  event.message.UtbildningstillfalleUID,
  {
    name: {
      sv: "Programmeringsteknik för F",
      en: "Programming in F",
    },
    courseCode: "DD1321",
    organisation: {
      code: "SCI",
    },
    organisationUnit: {
      code: "SCID",
      name: {
        sv: "Institutionen för datavetenskap",
        en: "School of Computer Science",
      },
    },
    courseInstanceCode: "DD1321-20221",
    startDate: "2023-01-17",
    endDate: "2023-03-17",
    credits: 7.5,
    modules: [
      {
        code: "F1",
        name: {
          sv: "Programmeringsteknik för F",
          en: "Programming in F",
        },
        credits: 7.5,
        gradingScheme: {
          code: "TH",
          grades: [
            {
              validFinalGrade: true,
              code: "P",
            },
            {
              validFinalGrade: false,
              code: "F",
            },
          ],
        },
      },
    ],
    gradingScheme: {
      code: "TH",
      grades: [
        {
          validFinalGrade: true,
          code: "P",
        },
        {
          validFinalGrade: false,
          code: "F",
        },
      ],
    },
  },
);

UgIntegrationMock.getUgCourseResponsibleAndTeachers("SF1625", "2022", "2", [
  "u1responsible",
  ["u1teacher1", "u1teacher2"],
]);

UgIntegrationMock.getUgUser("u1responsible", {
  email: "cr@email.com",
  kthid: "u1responsible",
  givenName: "Course",
  surname: "Responsible",
});

UgIntegrationMock.getUgUser("u1examiner", {
  email: "ex@email.com",
  kthid: "u1examiner",
  givenName: "Course",
  surname: "Examiner",
});

UgIntegrationMock.getUgUser("u1teacher1", {
  email: "t1@email.com",
  kthid: "u1teacher1",
  givenName: "First",
  surname: "Teacher",
});

UgIntegrationMock.getUgUser("u1teacher2", {
  email: "t2@email.com",
  kthid: "u1teacher2",
  givenName: "Second",
  surname: "Teacher",
});

UgIntegrationMock.getUgSchool("SCI", {
  name: "SCI",
  kthid: "u1school",
  description: {
    sv: "Skolan för datavetenskap och kommunikation",
    en: "School of Computer Science and Communication",
  },
});

describe("RegistreringEvent", () => {
  test("can be executed", async () => {
    const mockDb = new MockDatabase();
    const mockContext = new MockContext(event.userProps);
    await handler(event.message, mockContext, mockDb);
  });

  test("logs invocation", async () => {
    const mockDb = new MockDatabase();
    const mockContext = new MockContext(event.userProps);
    await handler(event.message, mockContext, mockDb);

    expect(mockContext.log.mock.calls.length).toBe(1);
    expect(mockContext.log.mock.lastCall[0]).toBe(
      `PaborjatUtbildningstillfalleEvent: ${event.message.UtbildningstillfalleUID}`,
    );
  });

  test.skip("fetches course info from KOPPS", async () => {
    const mockDb = new MockDatabase();
    const mockContext = new MockContext(event.userProps);
    await handler(event.message, mockContext, mockDb);
    // TODO: We meed to mock the KOPPS API
    expect(mockDb._result).toMatchSnapshot();
  });

  test.skip("fetches course info from LADOK", async () => {
    const mockDb = new MockDatabase();
    const mockContext = new MockContext(event.userProps);
    await handler(event.message, mockContext, mockDb);
    // TODO: We meed to mock the LADOK REST API
    expect(mockDb._result).toMatchSnapshot();
  });

  test.skip("fetches course info from UG", async () => {
    const mockDb = new MockDatabase();
    const mockContext = new MockContext(event.userProps);
    await handler(event.message, mockContext, mockDb);
    // TODO: We meed to mock the UG REST API
    expect(mockDb._result).toMatchSnapshot();
  });

  test("writes correct data to db", async () => {
    const mockDb = new MockDatabase();
    const mockContext = new MockContext(event.userProps);
    await handler(event.message, mockContext, mockDb);

    expect(mockDb._result).toMatchSnapshot();
  });
});
