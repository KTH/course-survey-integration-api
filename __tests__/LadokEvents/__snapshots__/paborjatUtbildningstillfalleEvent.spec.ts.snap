// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistreringEvent writes correct data to db 1`] = `
{
  "CourseRound": {
    "_gradingScheme": [
      "P",
      "F",
    ],
    "canceled": false,
    "canvasSisId": "41717c91-4028-11ee-bf53-2115569549a8",
    "courseCode": "DD1321",
    "courseExaminor": {
      "email": "dummy@email.com",
      "fullName": "Dummy User",
      "kthUserId": "u1dummy",
      "userName": "dummyuser",
    },
    "courseGoal": "The course aims to give basic knowledge of programming in F#.",
    "courseResponsible": {
      "email": "cr@email.com",
      "fullName": "Course Responsible",
      "kthUserId": "u1responsible",
      "userName": "cr",
    },
    "courseTeachers": [
      {
        "email": "t1@email.com",
        "fullName": "First Teacher",
        "kthUserId": "u1teacher1",
        "userName": "t1",
      },
      {
        "email": "t2@email.com",
        "fullName": "Second Teacher",
        "kthUserId": "u1teacher2",
        "userName": "t2",
      },
    ],
    "credits": "7.5",
    "displayYear": "2023",
    "endDate": "2023-03-17",
    "id": "41717c91-4028-11ee-bf53-2115569549a8",
    "institution": {
      "displayCode": "DUMMY",
      "displayName": "Dummy Institution",
      "kthId": "u1dummyinst",
    },
    "ladokCourseId": "e51b9586-9501-11ee-a0ce-a9a57d284dbd",
    "ladokCourseRoundId": "41717c91-4028-11ee-bf53-2115569549a8",
    "language": "sv",
    "modules": [
      {
        "_reportedResults": {},
        "code": "F1",
        "credits": 7.5,
        "gradingDistribution": {
          "F": -1,
          "P": -1,
        },
        "gradingScheme": [
          "P",
          "F",
        ],
        "name": "Programmeringsteknik för F",
        "nrofReportedResults": 0,
      },
    ],
    "name": "Programmeringsteknik för F",
    "organization": {
      "displayCode": "SCI",
      "displayName": "Skolan för datavetenskap och kommunikation",
      "kthId": "u1school",
    },
    "period": "P1",
  },
}
`;
